src/custom/constants/networks.ts

```diff
@@ -1,30 +1,146 @@
-  [SupportedChainId.MAINNET]: `https://mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.RINKEBY]: `https://rinkeby.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.ROPSTEN]: `https://ropsten.infura.io/v3/${INFURA_KEY}`,
-  [SupportedChainId.GOERLI]: `https://goerli.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.KOVAN]: `https://kovan.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.OPTIMISM]: `https://optimism-mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.OPTIMISTIC_KOVAN]: `https://optimism-kovan.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.ARBITRUM_ONE]: `https://arbitrum-mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.ARBITRUM_RINKEBY]: `https://arbitrum-rinkeby.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.POLYGON]: `https://polygon-mainnet.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.POLYGON_MUMBAI]: `https://polygon-mumbai.infura.io/v3/${INFURA_KEY}`,
-  // [SupportedChainId.CELO]: `https://forno.celo.org`,
-  // [SupportedChainId.CELO_ALFAJORES]: `https://alfajores-forno.celo-testnet.org`,
-  [SupportedChainId.GNOSIS_CHAIN]: `https://rpc.gnosis.gateway.fm`,
+class BaseUrlJsonRpcProvider extends UrlJsonRpcProvider {}
+type ProviderClass = Newable<typeof BaseUrlJsonRpcProvider>
+
+// List of public providers that don't require an API key, per chain support.
+const PUBLIC_PROVIDER_LIST: Record<SupportedChainId, ProviderClass[]> = {
+  [SupportedChainId.MAINNET]: [InfuraProvider, CloudflareProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GOERLI]: [InfuraProvider, AlchemyProvider, PocketProvider, AnkrProvider],
+  [SupportedChainId.GNOSIS_CHAIN]: [],
+}
+
+function getProvider(chainId: SupportedChainId): JsonRpcProvider[] {
+  const result: JsonRpcProvider[] = []
+
+  if (window.ethereum) {
+    result.push(new Web3Provider(window.ethereum, RPC_CONFIG[chainId]))
+  }
+
+  if (typeof RPC_CONFIG[chainId].rpcUrl === 'string') {
+    result.push(new StaticJsonRpcProvider(RPC_CONFIG[chainId].rpcUrl))
+  } else {
+    result.push(new InfuraProvider(RPC_CONFIG[chainId], INFURA_KEY))
+  }
+
+  if (result.length > 0) {
+    return result
+  }
+
+  for (const Provider of PUBLIC_PROVIDER_LIST[chainId]) {
+    try {
+      const provider = new Provider(RPC_CONFIG[chainId])
+
+      result.push(provider)
+    } catch (error) {
+      const { sentryError, tags } = constructSentryError(error, { message: "Couldn't create public provider" })
+
+      Sentry.captureException(sentryError, { tags })
+    }
+  }
+
+  return result
+}
+
+export const PROVIDERS: Record<SupportedChainId, JsonRpcProvider[]> = {
+  [SupportedChainId.MAINNET]: getProvider(SupportedChainId.MAINNET),
+  [SupportedChainId.GOERLI]: getProvider(SupportedChainId.GOERLI),
+  [SupportedChainId.GNOSIS_CHAIN]: getProvider(SupportedChainId.GNOSIS_CHAIN),
+}
+
+export const getRpcUrls = (chainId: SupportedChainId) => PROVIDERS[chainId].map((provider) => provider.connection.url)
```

Comment:
Why exporting this, plus the `consts`? I would think one thing or the other, right?

Reply:
This exports URLs specifically instead of the provider itself. So for example, when you are using the public provider, we can give you the URL. It is there to support wallets that expect a URL instead of a provider instance, while still allowing for a public provider to be injected.